proposals/p2173.md

```diff
@@ -0,0 +1,1355 @@
+convertible to) the type of the parameter", so long as the developer has some
+way to demonstrate that the types are the same if that is indeed the case.
+
+However, not all comparisons of generic types have that property. As an extreme
+and hopefully avoidable example, suppose a C++-like linkage model is used, with
+name mangling. We may need to determine whether two type expressions are equal
+at link time, by reducing them both to strings that are guaranteed to be
+identical if the types are equal and non-identical otherwise. This requires some
+form of type canonicalization: reducing all equal type expressions to some
+common representation where we can faithfully perform equality comparisons.
+
+Even if the semantics of Carbon don't require canonicalization, it is still a
+very useful property for implementation purposes. For example, if type
+canonicalization is possible to perform efficiently, building a data structure
+mapping from a type to some value becomes much simpler, and if the
+representation of a type expression carries its canonical form then type
+equality comparisons can be done in constant time.
+
+### Transitivity of equality
+
+For ergonomic purposes, it is valuable for the notion of type expression
+equality to be transitive: if type A is known to be the same as type B, and type
+B is known to be the same as type C, then shouldn't we be able to use an A where
+a C is required? See
+[#1369](https://github.com/carbon-language/carbon-lang/issues/1369) for an
+explanation of why this is important.
+
+However, if we permit arbitrary constraints and have a transitive equality rule,
+then type equality is isomorphic to the
+[word problem for finitely-presented monoids](https://en.wikipedia.org/wiki/Word_problem_for_groups),
+which is undecidable, and certainly not computable, in general.
+
+Therefore we must impose a restriction somewhere: either not all constraints are
+permissible, or we do not have transitivity, or we impose some other constraint
+on the language to avoid the hard cases. Some languages merely impose some kind
+of depth limit on their search for a proof of equality.
+
+Note that if we can perform type canonicalization, then type equality is
+necessarily transitive: if all equal _pairs_ of types canonicalize to the same
+representation, then _all_ equal types canonicalize to the same representation
+transitively.
+
+### Coherence
+
+In order to avoid surprises in the language behavior, we want Carbon to have the
+property that learning more about the type of a value should not cause the
+behavior of an expression to change from one valid behavior to another valid
+behavior.
+
+It's important to distinguish here between learning more about a type, such as
+learning that it implements a particular interface – for example, from an added
+import – and changing the type, such as by writing a different expression after
+a `:` or `:!`. Changing the type of a value is expected to be able to change the
```

Comment:
What I mean here is that changing from `a: i32` to `a: f32` changes the type of `a`, and changing from `T:! C` to `T:! C where .X = 5` changes the type of `T`. I think what you're observing is that changing from `T:! C = i32` to `T:! (C where .X = 5) = i32` doesn't change the type of `a: T` -- either way, it's `a: i32`, because the right-hand side of `:` is converted to `type`.

Is there some way I can update this paragraph to make it clearer what I'm talking about?

Reply:
I can't think of anything.

The root issue is "writing a different expression after a `:` or `:!`." -> we've just recently allowed that to *not* always change the type, so I get a little surprised when that is the first example of how to change the type of the value. But of course, it is, and different expressions that *don't* change the type of the value are the exception and not the rule. So I think its ultimately fine. =D